### ISSUE TITLE
uBO for Firefox connects to remote hosts without being explicitly asked to

### ISSUE THREAD

**Issue by anchev** [uBlockOrigin/uBlock-issues#38]
### Prerequisites

- [x] I verified that this is not a filter issue
- [x] This is not a support issue or a question
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
- I tried to reproduce the issue when...
    - [x] uBlock Origin is the only extension
    - [ ] uBlock Origin with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [x] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

uBO for Firefox connects to remote hosts without being explicitly asked to.

### A specific URL where the issue occurs

[not applicable]

### Steps to Reproduce

1. Download Firefox 60.0 x64 for Linux from firefox.com
2. (see attachment [settings.zip](https://github.com/uBlockOrigin/uBlock-issues/files/1996225/settings.zip)) Download `user.js` and save it in `/tmp/download`. This is pyllyukko's `user.js` with additional modifications to ensure zero packet sending without explicit user request.
3. Create new clean profile using the following script:
```
#!/bin/bash

PROFILENAME="myprofile"
PROFILEDIR="${PROFILENAME}.default"
PROFILE="$HOME/.mozilla/firefox/${PROFILEDIR}"
USERJS="/tmp/download/user.js"
# use path to your browser binary
BROWSER="/usr/local/bin/firefox/firefox" 

# Create clean profile
rm -rf ${PROFILE}
$BROWSER -CreateProfile "${PROFILENAME} ${PROFILE}"

# Create a symlink to the downloaded user.js
ln -s $USERJS "${PROFILE}/user.js"

# Start browser
$BROWSER -P "${PROFILENAME}" --no-remote about:blank

exit
```
4. Install uBlock Origin
5. Go to `about:addons`->Extensions and turn off automatic updates
6. Go to uBO's preferences and "Restore from file..." attached settings.

Note: only Built-in filters are enabled (except the experimantal one) and "Auto-update filter lists" is turned off. Additionally the `behind-the-scene` rule is commented in the "Whitelist" section

7. Click "Update now" to update filter lists
8. Exit firefox
9. Stop any system network services (such as ntpd or similar).
10. Run in console:
```
rcnetwork restart;tcpdump -i eth1 ip src host pc and dst host not router and dst host not pc -tq
```
11. Start firefox and watch console output

### Expected behavior:

Zero network packets, no connections whatsoever. (This is how it works without installing uBlock Origin)

### Actual behavior:

`tcpdump` output:
```
IP pc.46954 > 151.101.0.133.https: tcp 0
IP pc.46954 > 151.101.0.133.https: tcp 0
IP pc.46954 > 151.101.0.133.https: tcp 517
IP pc.46954 > 151.101.0.133.https: tcp 0
IP pc.46954 > 151.101.0.133.https: tcp 93
IP pc.46954 > 151.101.0.133.https: tcp 335
IP pc.46954 > 151.101.0.133.https: tcp 0
IP pc.46954 > 151.101.0.133.https: tcp 0
IP pc.46954 > 151.101.0.133.https: tcp 31
IP pc.46954 > 151.101.0.133.https: tcp 0
IP pc.46954 > 151.101.0.133.https: tcp 0
IP pc.46954 > 151.101.0.133.https: tcp 0
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 191
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 126
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 364
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 203
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 126
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 379
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 31
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.38272 > ec2-54-69-217-61.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 31
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
IP pc.33192 > ec2-35-166-131-171.us-west-2.compute.amazonaws.com.https: tcp 0
```

### Your environment

* uBlock Origin version: 1.16.4
* Browser Name and version: Firefox 60.0 (64-bit) for Linux
* Operating System and version: openSUSE Leap 42.3

**Comment by uBlock-user** [comment]
Is  `Auto-update filter lists` ticked ?

**Comment by anchev** [comment]
> Is Auto-update filter lists ticked ?

No. This is mentioned in the note after step 6. You also have it in the attached settings.

**Comment by gorhill** [comment]
Are there any filter list selected which are not part of the package and which are not in uBO's cache (or is uBO prevented from caching)?

uBO never ever connect to remove server without auto-update or without user expressly clicking "Update now", **except** if one or more selected filter lists are not part of the package and not in uBO's cache.

**Comment by gorhill** [comment]
Instead of cryptic TCP dump output, please just look at the browser console to find out exactly where the browser connect, no need for external tools. Use the Network pane the browser console, which can be opened automatically with the `-jsconsole` flag.

**Comment by gorhill** [comment]
You have `ublock-annoyances` selected as filter list, that list is not part of uBO's package. So the remote connection would be necessary for uBO to fetch the list if it's not in its cache.

**Comment by gorhill** [comment]
I create a new profile and installed `https://github.com/pyllyukko/user.js`, and this caused the setting _"Always use private browsing mode"_ to be enabled.

That setting prevents uBO from caching the fetched filter lists, meaning it will have to fetch them at every re-start of Firefox: <https://bugzilla.mozilla.org/show_bug.cgi?id=1313401>.

Duplicate of <https://github.com/gorhill/uBlock/issues/3071>.

**Comment by anchev** [comment]
Good notice about private browsing. Indeed I tested setting:
```
user_pref("browser.privatebrowsing.autostart", false);
```
in `user.js` and updating the filters results in no connections.

So yet another case we need to thank Mozilla for their "privacy respecting" browser which denies the possibility of privacy, especially in "private browsing mode". Now you know why I prefer to test with external independent tool rather than any built-in browser consoles or whatever :)

**Comment by anchev** [comment]
BTW with uMatrix this doesn't happen (tcpdump shows no connections, even with `user_pref("browser.privatebrowsing.autostart", true);`). May I ask why?

**Comment by gorhill** [comment]
In uMatrix, all the hosts files are part of the package. Add an external hosts file and this will happen.

**Comment by uBlock-user** [comment]
Be well aware before installing any _user.js_, otherwise you will tend to run into such issues.

**Comment by anchev** [comment]
> @anchev you are asking to repeat what I already said.

Not at all. I am just trying to understand what "part of the package" means. To my mind "part of the package" as "included inside" and "no external dependencies".

> In uMatrix, all the hosts files are part of the package

> If you want to ascertain what filter lists/hosts files are packaged in the extensions, just go to the extension installations and see for yourself.

Alright. Here are the steps I followed:

- Instead of installing uBO (as explained above) I installed uMatrix only.
- Just as with uBO I turned off automatic updates of the extension and "Auto-update assets".
- I did *not* click "Update now".
- Exited firefox.
- Started firefox.

**Result:** no connections whatsoever (with or without persistent private browsing mode).

- Tried to find what is included in the package by searching:
```
~/.mozilla/firefox/myprofile.default]: grep -ir pollock .
Binary file ./storage/default/moz-extension+++c0286891-a750-4336-99aa-939a2f6cdc4b/idb/1576378395ueMgaatrroitxSCeahc.sqlite matches
./browser-extension-data/uMatrix@raymondhill.net/storage.js:{"selectedHostsFiles":["malware-0","malware-1","dpollock-0","hphosts","mvps-0","plowe-0"],"selectedRecipeFiles":[],"liveHostsFiles":[["malware-0",{"title":"Malware Domain List","contentURL":["https://www.malwaredomainlist.com/hostslist/hosts.txt","assets/thirdparties/www.malwaredomainlist.com/hostslist/hosts.txt"],"type":"filters","hasLocalURL":true,"hasRemoteURL":true,"updateAfter":13,"writeTime":0,"obsolete":true,"selected":true,"entryCount":1137,"entryUsedCount":1137}],["malware-1",{"title":"Malware domains","contentURL":["https://mirror.cedia.org.ec/malwaredomains/justdomains","https://mirror1.malwaredomains.com/files/justdomains","assets/thirdparties/mirror1.malwaredomains.com/files/justdomains","assets/thirdparties/mirror1.malwaredomains.com/files/justdomains.txt"],"supportURL":"http://www.malwaredomains.com/","type":"filters","hasLocalURL":true,"hasRemoteURL":true,"updateAfter":13,"writeTime":0,"obsolete":true,"selected":true,"entryCount":19794,"entryUsedCount":19779}],["dpollock-0",{"updateAfter":11,"title":"Dan Pollock’s hosts file","contentURL":["http://someonewhocares.org/hosts/hosts","assets/thirdparties/someonewhocares.org/hosts/hosts.txt"],"supportURL":"http://someonewhocares.org/hosts/","type":"filters","hasLocalURL":true,"hasRemoteURL":true,"writeTime":0,"obsolete":true,"selected":true,"entryCount":13407,"entryUsedCount":13392}],["hphosts",{"updateAfter":11,"title":"hpHosts’ Ad and tracking servers","contentURL":["https://hosts-file.net/.%5Cad_servers.txt","assets/thirdparties/hosts-file.net/ad_servers.txt"],"supportURL":"https://hosts-file.net/","type":"filters","hasLocalURL":true,"hasRemoteURL":true,"writeTime":0,"obsolete":true,"selected":true,"entryCount":45728,"entryUsedCount":40391}],["mvps-0",{"updateAfter":11,"title":"MVPS HOSTS","contentURL":["http://winhelp2002.mvps.org/hosts.txt","assets/thirdparties/winhelp2002.mvps.org/hosts.txt"],"supportURL":"http://winhelp2002.mvps.org/","type":"filters","hasLocalURL":true,"hasRemoteURL":true,"writeTime":0,"obsolete":true,"selected":true,"entryCount":12714,"entryUsedCount":10812}],["plowe-0",{"updateAfter":13,"title":"Peter Lowe’s Ad and tracking server list","contentURL":["https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts&showintro=1&mimetype=plaintext","assets/thirdparties/pgl.yoyo.org/as/serverlist","assets/thirdparties/pgl.yoyo.org/as/serverlist.txt"],"supportURL":"https://pgl.yoyo.org/adservers/","type":"filters","hasLocalURL":true,"hasRemoteURL":true,"writeTime":0,"obsolete":true,"selected":true,"entryCount":2647,"entryUsedCount":1823}]],"alwaysDetachLogger":false,"autoUpdate":false,"clearBrowserCache":true,"clearBrowserCacheAfter":60,"cloudStorageEnabled":false,"collapseBlacklisted":true,"collapseBlocked":false,"colorBlindFriendly":false,"deleteCookies":false,"deleteUnusedSessionCookies":false,"deleteUnusedSessionCookiesAfter":60,"deleteLocalStorage":false,"displayTextSize":"14px","externalHostsFiles":[],"externalRecipeFiles":[],"iconBadgeEnabled":true,"maxLoggedRequests":1000,"noTooltips":false,"popupCollapseAllDomains":false,"popupCollapseBlacklistedDomains":false,"popupScopeLevel":"domain","processHyperlinkAuditing":true,"processReferer":false,"userHosts":{"enabled":false,"content":""},"userRecipes":{"enabled":false,"content":""}}
```
Checking the size of the sqlite file which grep found:
```
-rw------- 1 xxx users 147456 May 12 15:07 1576378395ueMgaatrroitxSCeahc.sqlite
```
Obviously quite smaller than the 412Kb of the original http://someonewhocares.org/hosts/hosts. Also nothing seems to contain the string `goatse.cx` which is part of the hosts file:
```
[~/.mozilla/firefox/myprofile.default]: grep -ir "goatse.cx" .
```
Anyway I went ahead and opened the DB using sqlitebrowser. The tables were pretty much empty. Yet attempting to open `goatse.cx` in the browser showed a warning that uMatrix prevented loading it. So obviously somehow uMatrix knows about this particular domain, without updating. One could assume it is "part of the package".

But after clicking "Update now" the sqlite file became more than 7Mb which implies that obviously not all the info is included as "part of the package". Also the very fact that this update function exists means downloading of external data.

Hence the logical question: what is "part of the package"? Is it some basic short list that is included (but quite dated and needing much more data which is not part of the package to be fully functional)? If yes: Does it mean that it is only this "basic" and "part of the package" list is used when "Always use private browsing mode" is enabled? And why is that not the case for uBO?

That's why I asked explicitly about uBO's Built-in list "uBlock filters" (which I see is part of the .xpi in `assets/ublock/filters.txt` and supposedly "part of the package"): How come having *only it* enabled causes connections (in relation to Firefox's bug you showed) unlike uMatrix? In short: why the two extensions behave differently under the same testing conditions? If it is only Firefox to blame and being "part of the package" means that no connection should be made when "Auto update" is turned off - then this is a bug for uBO. Or if connections should be made - then this is a bug for uMatrix.

I hope you understand the confusion.

**Comment by gorhill** [comment]
> How come having only it enabled causes connections

It does not, browser console shows all network requests done at start-up (using `-jsconsole`):

![a](https://user-images.githubusercontent.com/585534/39957911-c43c85b2-55c8-11e8-927e-b2cd72bb7e23.png)

In your original report you had "uBlock filters -- Annoyances" selected. This list is not in the package because it is not selected by default.

**Comment by gorhill** [comment]
For completion, here is the result with selecting a filter list which is not part of the package:

![a](https://user-images.githubusercontent.com/585534/39957969-d18f4b72-55c9-11e8-90d7-55d692915a04.png)

As seen, the browser console reports just fine what uBO fetched, so no need to presume it does not work as intended.

**Comment by anchev** [comment]
OK. I tested again. You are right. Thanks for explaining.

Perhaps in this case "annoyances" needs to be moved out of the "Built-in" lists if it is not actually built in.

**Comment by gorhill** [comment]
"Built-in" is to express that the lists are uBO-maintained, not that they are necessarily part of the package. See <https://github.com/gorhill/uBlock/issues/3708>.

**Comment by anchev** [comment]
Then maybe some other indication of what is part of the package would be good to avoid confusion as in this issue. Maybe a custom icon with a link to a short wiki page explaining what we discussed here.

**Comment by gorhill** [comment]
> Then maybe some other indication of what is part of the package

That is not really important -- I chose to include filter lists which are default. But then uBO will select by default extra filter lists based on browser language, and these are not part of the package. I don't know why you focus so much on such a implementation details. As said above, these uBO lists must be seen as complementary to EasyList/EasyPrivacy, using them on their own is not going to accomplish much, selecting them or not should not be because they are currently part of the package or not. 

As far as I am concerned I have spent quite an amount of time on this non-issue, so I am bailing out.
